service/s3/handwritten-paginators.go

```diff
@@ -0,0 +1,193 @@
+package s3
+
+import (
+	"context"
+	"fmt"
+)
+
+// ListObjectVersionsAPIClient is a client that implements the ListObjectVersions
+// operation
+type ListObjectVersionsAPIClient interface {
+	ListObjectVersions(context.Context, *ListObjectVersionsInput, ...func(*Options)) (*ListObjectVersionsOutput, error)
+}
+
+var _ ListObjectVersionsAPIClient = (*Client)(nil)
+
+// ListObjectVersionsPaginatorOptions is the paginator options for ListObjectVersions
+type ListObjectVersionsPaginatorOptions struct {
+	// (Optional) The maximum number of Object Versions that you want Amazon Route 53 to
+	// return.
+	Limit int32
+
+	// Set to true if pagination should stop if the service returns a pagination token
+	// that matches the most recent token provided to the service.
+	StopOnDuplicateToken bool
+}
+
+// ListObjectVersionsPaginator is a paginator for ListObjectVersions
+type ListObjectVersionsPaginator struct {
+	options         ListObjectVersionsPaginatorOptions
+	client          ListObjectVersionsAPIClient
+	params          *ListObjectVersionsInput
+	firstPage       bool
+	KeyMarker       *string
+	VersionIdMarker *string
+}
+
+// NewListObjectVersionsPaginator returns a new ListObjectVersionsPaginator
+func NewListObjectVersionsPaginator(client ListObjectVersionsAPIClient, params *ListObjectVersionsInput, optFns ...func(*ListObjectVersionsPaginatorOptions)) *ListObjectVersionsPaginator {
+	if params == nil {
+		params = &ListObjectVersionsInput{}
+	}
+
+	options := ListObjectVersionsPaginatorOptions{}
+	options.Limit = params.MaxKeys
+
+	for _, fn := range optFns {
+		fn(&options)
+	}
+
+	return &ListObjectVersionsPaginator{
+		options:         options,
+		client:          client,
+		params:          params,
+		firstPage:       true,
+		KeyMarker:       params.KeyMarker,
+		VersionIdMarker: params.VersionIdMarker,
+	}
+}
+
+// HasMorePages returns a boolean indicating whether more pages are available
+func (p *ListObjectVersionsPaginator) HasMorePages() bool {
+	return p.firstPage || (p.KeyMarker != nil && len(*p.KeyMarker) != 0)
```

Comment:
I don't think so. 
If `keyMarker` is not nil, then there has to be a `versionIdMarker` associated with it, even if it only has 1 version.

I can add that field to the validation as a "just in case"

```go
func (p *ListObjectVersionsPaginator) HasMorePages() bool {
	return p.firstPage || (p.keyMarker != nil && len(*p.keyMarker) != 0) && (p.versionIdMarker != nil && len(*p.versionIdMarker) != 0)
}

```

Reply:
I would take a look at one of the SDKs that already has these paginators (e.g. Java) to understand the correct behavior.

Reply:
[Here](https://code.amazon.com/packages/AwsJavaSdk-S3/blobs/c2701ce6de494f74a28d1fdaa9bbc4c355fc74c0/--/src/software/amazon/awssdk/services/s3/paginators/ListMultipartUploadsIterable.java#L148) is where this is determined in the Java SDK. It only looks at `isTruncated` to determine if there are more pages. 
I think it makes sense and is more elegant.